src/ips.rs

```diff
@@ -0,0 +1,525 @@
+            // nodes should pursue to degree_delta level. That could be bad if graph's vertexes
+            // have very high (or low) degrees and therefore, delta is very high (or low) too. But until
+            // we have some better idea this one is the best we can do to keep up with the graph.
+            let desired_degree = (degree_avg as u32 + degrees.get(&node_ip).unwrap()) / 2;
+
+            // 3 - Calculate how many peers to add or delete from peerlist
+            let peers_to_delete_count = if desired_degree < *degrees.get(&node_ip).unwrap() {
+                degrees.get(&node_ip).unwrap() - desired_degree
+            } else {
+                // Check if config forces to change peerlist even if we have good degree.
+                // This should be always set to at least one to allow for some changes in graph -
+                // searching for better potential peers.
+                self.config.change_at_least
+            };
+
+            // Calculating how many peers should be added. If we have more peers than desired degree
+            // we will add at least config.change_at_least peers.
+            let mut peers_to_add_count = if desired_degree > *degrees.get(&node_ip).unwrap() {
+                desired_degree - degrees.get(&node_ip).unwrap() + peers_to_delete_count
+            } else {
+                self.config.change_at_least
+            };
+
+            // Limit number of changes to config value
+            if peers_to_add_count > self.config.change_no_more {
+                peers_to_add_count = self.config.change_no_more;
+            }
+
+            // 4 - Choose peers to delete from peerlist (based on ranking)
+            if peers_to_delete_count > 0 {
+                // Sort peers by rating (highest first)
+                curr_peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
+
+                // Remove peers with lowest rating
+                for _ in 0..peers_to_delete_count {
+                    let peer_to_delete = curr_peer_ratings.pop();
+                    if let Some(peer_to_delete) = peer_to_delete {
+                        peer_list_entry.list.retain(|x| x != &peer_to_delete.0);
+                    }
+                }
+            }
+
+            // 5 - Find peers to add from selected peers (based on rating)
+            if peers_to_add_count > 0 {
+                // Take twice as many candidates as we need to add to peerlist to be able to
+                // choose best ones from them.
+                let mut candidates_to_search = peers_to_add_count * 2;
+
+                let mut candidates: Vec<(IpAddr, usize, f64)> =
+                    Vec::with_capacity(candidates_to_search as usize);
+
+                // Sort peers by rating
+                peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
+
+                // Add peers with highest rating as candidates
+                for peer in peer_ratings.iter() {
```

Comment:
I think this can perhaps be written more cleanly with iterators? i.e. with `filter` to check for inclusion and then perhaps a `collect`? Just a thought...